scipy/sparse/linalg/_isolve/utils.py

```diff
@@ -79,6 +80,10 @@ def make_system(A, M, x0, b):
         b = b.astype('d')  # upcast non-FP types to double
 
     def postprocess(x):
+        if x0shape is None:
```

Comment:
`x0` is optional. Solution should follow `b` shape and not `x0`. `b` is always given and should override initial guess even when it is given. It is OK to ravel `x0` at all times. It has no bearing on the solution.

Reply:
It doesn't matter if x0 is ravel internally.

What matters is the expectation of the user on the output. As a user, I expect the solution to be of the same shape as my initial guess if given.

Reply:
It does internally that we don't get strange things internally and spend extra if then else in the solvers polluting logic. 

I don't agree with the initial guess. It can come in any shape. What matters is the RHS. Guess is an optional input. Nothing to do with the solution algorithm.

Reply:
@ilayn Please see https://github.com/scipy/scipy/pull/21164#issuecomment-2227496340 and react